### PR TITLE
ENG-8819 Inline limit for max

### PR DESCRIPTION
**Pull request by yzhang1991**

### CODE REVIEW

**Comment by yzhang1991** [comment]
This is just a duplicate of ENG-4016 pull request, because we created a new ticket for this optimization.

**Comment by yzhang1991** [comment]
Need to add more test cases for parameterized plans.

File: src/frontend/org/voltdb/planner/microoptimizations/ReplaceWithIndexLimit.java [hunk 1]
```diff
@@ -208,34 +208,88 @@ protected AbstractPlanNode recursivelyApply(AbstractPlanNode plan)
...
+                e.printStackTrace();
+                assert(false);
                 return plan;
             }
         }
 
+        /* If there is only 1 difference between searchkeyExprs and endExprs,
+         * 1. trivial filters can be discarded, 2 possibilities:
+         *      a. SELECT MIN(X) FROM T WHERE [other prefix filters] X < / <= ?
+         *         <=> SELECT MIN(X) FROM T WHERE [other prefix filters] && the X < / <= ? filter
```
Comment (paulmartel) [inline src/frontend/org/voltdb/planner/microoptimizations/ReplaceWithIndexLimit.java]: I don't understand what "<=>" means here. Also, should " WHERE [other prefix filters] X < / <= ? " be " WHERE [other prefix filters AND] X < / <= ? " ?
What does "&&" mean here? What's the difference supposed to be between the query before the <=> and the one after?

File: src/frontend/org/voltdb/planner/microoptimizations/ReplaceWithIndexLimit.java [hunk 2]
```diff
@@ -208,34 +208,88 @@ protected AbstractPlanNode recursivelyApply(AbstractPlanNode plan)
...
+             * X < ? will be a post filter for the scan rather than an initial expression. */
+            if (numberOfExprs == 1) {
+                // e.g.: explain select max(c2/2) from t where c2/2<=3;
+                // In this case, as long as the where condition (exprs.get(0)) matches the aggregation argument, continue.
+                AbstractExpression exprToBind = indexedExprs == null ? exprs.get(0).getLeft() : indexedExprs.get(0);
+                if ( aggExpr.bindingToIndexedExpression(exprToBind) == null ) {
+                    return plan;
+                }
+            }
+            if (numberOfExprs > 1) {
```
Comment (paulmartel) [inline src/frontend/org/voltdb/planner/microoptimizations/ReplaceWithIndexLimit.java]: ELSE if

File: src/frontend/org/voltdb/planner/microoptimizations/ReplaceWithIndexLimit.java [hunk 3]
```diff
@@ -208,34 +208,88 @@ protected AbstractPlanNode recursivelyApply(AbstractPlanNode plan)
...
+                // ENG-4016: Optimization for query SELECT MAX(X) FROM T WHERE [other prefix filters] X < / <= ?
+                // Just keep trying, don't return early.
+                boolean earlyReturn = true;
+                for (int i=0; i<exprs.size(); ++i) {
```
Comment (paulmartel) [inline src/frontend/org/voltdb/planner/microoptimizations/ReplaceWithIndexLimit.java]: Use numberOfExprs here in place of exprs.size().

File: tests/frontend/org/voltdb/regressionsuites/TestIndexLimitSuite.java [hunk 4]
```diff
@@ -57,7 +57,13 @@ void callWithExpectedResult(Client client, int ret, String procName, Object... p
         VoltTable result = cr.getResults()[0];
         assertEquals(1, result.getRowCount());
         assertTrue(result.advanceRow());
-        assertEquals(ret, result.getLong(0));
+        if (ret == null) {
+            result.getLong(0);
+            assertTrue(result.wasNull());
```
Comment (paulmartel) [inline tests/frontend/org/voltdb/regressionsuites/TestIndexLimitSuite.java]: Give better diagnostics:

    long unexpected = result.getLong(0);
    assertTrue("Unexpected non-null value: " + unexpected, result.wasNull());

File: tests/frontend/org/voltdb/regressionsuites/TestIndexLimitSuite.java [hunk 5]
```diff
@@ -169,6 +175,12 @@ public void testSpecialCases() throws Exception {
         callWithExpectedResult(client, 3, "EDGE_TU2_MAX_POINTS1", "thea", 1);
         callWithExpectedResult(client, -1, "EDGE_TU2_MIN_POINTS2", 5, "jim");
         callWithExpectedResult(client, 3, "EDGE_TU2_MAX_POINTS2", 1, "thea");
+        callWithExpectedResult(client, 2, "EDGE_TU2_MAX_POINTS3", 3, "jim");
+        callWithExpectedResult(client, 3, "EDGE_TU2_MAX_POINTS4", 3, "jim");
+        callWithExpectedResult(client, (Integer)null, "EDGE_TU2_MAX_POINTS3", -1, "jim");
```
Comment (paulmartel) [inline tests/frontend/org/voltdb/regressionsuites/TestIndexLimitSuite.java]: reorder the test cases to put the ...POINTS3 calls together.

File: tests/frontend/org/voltdb/regressionsuites/TestIndexLimitSuite.java [hunk 6]
```diff
@@ -393,6 +405,10 @@ static public Test suite() {
         project.addStmtProcedure("EDGE_TU2_MAX_POINTS1", "SELECT MAX(POINTS) FROM TU2 WHERE UNAME = ? AND POINTS > ?");
         project.addStmtProcedure("EDGE_TU2_MIN_POINTS2", "SELECT MIN(POINTS) FROM TU2 WHERE POINTS < ? AND UNAME = ?");
         project.addStmtProcedure("EDGE_TU2_MAX_POINTS2", "SELECT MAX(POINTS) FROM TU2 WHERE POINTS > ? AND UNAME = ?");
+        project.addStmtProcedure("EDGE_TU2_MAX_POINTS3", "SELECT MAX(POINTS) FROM TU2 WHERE POINTS < ? AND UNAME = ?");
+        project.addStmtProcedure("EDGE_TU2_MAX_POINTS4", "SELECT MAX(POINTS) FROM TU2 WHERE POINTS <= ? AND UNAME = ?");
+        project.addStmtProcedure("EDGE_TU2_MAX_POINTS5", "SELECT MAX(POINTS) FROM TU2 WHERE POINTS = ? AND UNAME = ?");
+        project.addStmtProcedure("EDGE_TU2_MAX_POINTS6", "SELECT MAX(POINTS) FROM TU2 WHERE POINTS IS NULL AND UNAME = ?");
```
Comment (paulmartel) [inline tests/frontend/org/voltdb/regressionsuites/TestIndexLimitSuite.java]: For completeness, we should add all the MIN cases that match these new MAX cases.

File: tests/frontend/org/voltdb/regressionsuites/TestIndexLimitSuite.java (hunk 5 above)
Comment (paulmartel) [inline tests/frontend/org/voltdb/regressionsuites/TestIndexLimitSuite.java]: test POINTS3 and POINTS4 w/ 4, "jim"  to cover strict less than case.

File: src/frontend/org/voltdb/planner/microoptimizations/ReplaceWithIndexLimit.java [hunk 7]
```diff
@@ -208,34 +208,88 @@ protected AbstractPlanNode recursivelyApply(AbstractPlanNode plan)
...
-                    (numberOfExprs == 1 && aggExpr.bindingToIndexedExpression(exprs.get(0).getLeft()) == null)) {
+        /* Retrieve the index expressions from the target index. (ENG-8819, Ethan)
+         * This is because we found that for the following two queries:
+         *     #1: explain select max(c2/2) from t where c1=1 and c2/2<=3;
+         *     #2: explain select max(c2/2) from t where c1=1 and c2/2<=?;
+         * We can get an inline limit 1 for #2 but not for #1. This is because all constants in #1 got parameterized.
+         * The result is that the query cannot pass the bindingToIndexedExpression() tests below
+         * because we lost all the constant value expressions (cannot attempt to bind a pve to a pve!).
+         * Those constant values expressions can only be accessed from the idnex.
+         * We will not add those bindings to the ispn.getBindings() here because they will be added anyway in checkIndex().
```
Comment (paulmartel) [inline src/frontend/org/voltdb/planner/microoptimizations/ReplaceWithIndexLimit.java]: I don't believe this is true. If I plan:

    select max(c2/2) from t where c1=1 and c2/2<=3;

and then

    select max(c2/-3) from t where c1=1 and c2/2<=3;

do I NOT try to reuse the cached plan?

**Comment by paulmartel** [comment]
End of review round.

File: src/frontend/org/voltdb/planner/microoptimizations/ReplaceWithIndexLimit.java (hunk 1 above)
Comment (yzhang1991) [inline src/frontend/org/voltdb/planner/microoptimizations/ReplaceWithIndexLimit.java]: This paragraph of comment is not added by me, it's moved here from top. So I don't quite understand.

**Pull request closed without merging**